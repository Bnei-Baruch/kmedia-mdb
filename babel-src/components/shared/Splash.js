'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorSplash = exports.FrownSplash = exports.LoadingSplash = exports.Splash = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _semanticUiReact = require('semantic-ui-react');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Splash = exports.Splash = function Splash(props) {
  var text = props.text,
      subtext = props.subtext,
      icon = props.icon,
      color = props.color,
      isLoading = props.isLoading;


  return _react2.default.createElement(
    _semanticUiReact.Header,
    { as: 'h2', icon: true, textAlign: 'center' },
    _react2.default.createElement(_semanticUiReact.Icon, { name: icon, loading: isLoading, color: color }),
    _react2.default.createElement(
      _semanticUiReact.Header.Content,
      null,
      text,
      subtext ? _react2.default.createElement(
        _semanticUiReact.Header.Subheader,
        null,
        subtext
      ) : null
    )
  );
};

Splash.propTypes = {
  icon: _propTypes2.default.string.isRequired,
  text: _propTypes2.default.node.isRequired,
  subtext: _propTypes2.default.node,
  color: _propTypes2.default.string,
  isLoading: _propTypes2.default.bool
};

Splash.defaultProps = {
  text: '',
  subtext: null,
  color: 'black',
  isLoading: false
};

var LoadingSplash = exports.LoadingSplash = function LoadingSplash(props) {
  return _react2.default.createElement(Splash, Object.assign({ isLoading: true, icon: 'spinner' }, props));
};
var FrownSplash = exports.FrownSplash = function FrownSplash(props) {
  return _react2.default.createElement(Splash, Object.assign({ icon: 'frown', color: 'orange' }, props));
};
var ErrorSplash = exports.ErrorSplash = function ErrorSplash(props) {
  return _react2.default.createElement(Splash, Object.assign({ icon: 'warning sign', color: 'red' }, props));
};