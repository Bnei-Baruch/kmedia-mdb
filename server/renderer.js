import { URL } from 'url';
import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { matchRoutes } from 'react-router-config';
import { createMemoryHistory } from 'history';
import pick from 'lodash/pick';
import moment from 'moment';
import qs from 'qs';
import serialize from 'serialize-javascript';
import UAParser from 'ua-parser-js';
import localStorage from 'mock-local-storage';
import { parse as cookieParse } from 'cookie';
import { HelmetProvider } from 'react-helmet-async';

import routes from '../src/routes';
import { COOKIE_CONTENT_LANG, LANG_UI_LANGUAGES, LANG_UKRAINIAN } from '../src/helpers/consts';
import { getLanguageDirection, getLanguageLocaleWORegion } from '../src/helpers/i18n-utils';
import { getLanguageFromPath } from '../src/helpers/url';
import { isEmpty } from '../src/helpers/utils';
import createStore from '../src/redux/createStore';
import { actions as ssr } from '../src/redux/modules/ssr';
import { actions as settings } from '../src/redux/modules/settings';
import App from '../src/components/App/App';
import i18nnext from './i18nnext';

const helmetContext = {};

const DoNotRemove = localStorage; // DO NOT REMOVE - the import above does all the work

const BASE_URL = process.env.REACT_APP_BASE_URL;

// see https://yoast.com/rel-canonical/
function canonicalLink(req, lang) {
  // strip ui language from path
  let cPath = req.originalUrl;
  if (lang && cPath.startsWith(`/${lang}`)) {
    cPath = cPath.substring(3);
  }

  const s = cPath.split('?');

  // start with path part
  cPath = s[0];

  // strip leading slash as it comes from BASE_URL
  if (cPath.startsWith('/')) {
    cPath = cPath.substring(1);
  }

  // strip trailing slash as we don't use those
  if (cPath.endsWith('/')) {
    cPath = cPath.substring(0, cPath.length - 1);
  }

  // clean content invariant keys from search part (query)
  if (s.length > 1) {
    const q = qs.parse(s[1]);
    delete q.sstart;      // player slice start
    delete q.send;        // player slice end
    delete q.language;    // content language

    // This one is conceptually wrong.
    // As changing the active part in a playlist would most definitely
    // change the page's content.
    // Most of the content would seem to duplicate the CU page.
    // However, it's logically not easy and some units may not
    // have their own CU page so we skip it.
    delete q.ap;          // playlist active part

    if (!isEmpty(q)) {
      cPath = `${cPath}?${qs.stringify(q)}`;
    }
  }

  return `<link rel="canonical" href="${BASE_URL}${cPath}" />`;
}

// see https://yoast.com/hreflang-ultimate-guide/
function alternateLinks(req, lang) {
  // strip ui language from path
  let aPath = req.originalUrl;
  if (lang && aPath.startsWith(`/${lang}`)) {
    aPath = aPath.substring(3);
  }

  // strip leading slash as it comes from BASE_URL
  if (aPath.startsWith('/')) {
    aPath = aPath.substring(1);
  }

  return LANG_UI_LANGUAGES
    .map((x) => {
      const l = getLanguageLocaleWORegion(x);
      return `<link rel="alternate" href="${BASE_URL}${x}/${aPath}" hreflang="${l}" />`;
    })
    .join('');
}

function ogUrl(req, lang) {
  // strip ui language from path
  let aPath = req.originalUrl;
  if (lang && aPath.startsWith(`/${lang}`)) {
    aPath = aPath.substring(3);
  }

  // strip leading slash as it comes from BASE_URL
  if (aPath.startsWith('/')) {
    aPath = aPath.substring(1);
  }

  return `<meta property="og:url" content="${BASE_URL}${lang}/${aPath}" />`;
}

export default function serverRender(req, res, next, htmlData) {
  console.log('serverRender', req.originalUrl);

  const { language, redirect } = getLanguageFromPath(req.originalUrl, req.headers);
  if (redirect) {
    const newUrl = `${BASE_URL}${language}${req.originalUrl}`;
    console.log(`serverRender: redirect (${language}) => ${newUrl}`);
    res.writeHead(307, { Location: newUrl });
    res.end();
    return;
  }

  moment.locale(language === LANG_UKRAINIAN ? 'uk' : language);

  const i18nServer = i18nnext.cloneInstance();
  i18nServer.changeLanguage(language, (err) => {
    if (err) {
      next(err);
      return;
    }

    const history = createMemoryHistory({
      initialEntries: [req.originalUrl],
    });

    const initialState = {
      router: { location: history.location },
      device: { deviceInfo: new UAParser(req.get('user-agent')).getResult() },
    };

    const store = createStore(initialState, history);

    const context = {
      req,
      data: {},
      head: [],
      i18n: i18nServer,
    };

    const reqPath = req.originalUrl.split('?')[0];
    const branch  = matchRoutes(routes, reqPath);

    let hrstart    = process.hrtime();
    const promises = branch.map(({ route, match }) => (
      route.ssrData
        ? route.ssrData(store, { ...match, parsedURL: new URL(req.originalUrl, 'https://example.com') })
        : Promise.resolve(null)
    ));

    let hrend = process.hrtime(hrstart);
    console.log('serverRender: fire ssrLoaders %ds %dms', hrend[0], hrend[1] / 1000000);
    hrstart = process.hrtime();

    Promise.all(promises)
      .then(() => {
        store.stopSagas();
        hrend = process.hrtime(hrstart);
        console.log('serverRender: Promise.all(promises) %ds %dms', hrend[0], hrend[1] / 1000000);
        hrstart = process.hrtime();

        store.rootSagaPromise
          .then(() => {
            hrend = process.hrtime(hrstart);
            console.log('serverRender: rootSagaPromise.then %ds %dms', hrend[0], hrend[1] / 1000000);
            hrstart = process.hrtime();

            // actual render
            const markup = ReactDOMServer.renderToString(<HelmetProvider context={helmetContext}><App i18n={context.i18n} store={store} history={history} /></HelmetProvider>);
            hrend        = process.hrtime(hrstart);
            console.log('serverRender: renderToString %ds %dms', hrend[0], hrend[1] / 1000000);
            hrstart = process.hrtime();

            const { helmet } = helmetContext;
            hrend            = process.hrtime(hrstart);
            console.log('serverRender:  Helmet.renderStatic %ds %dms', hrend[0], hrend[1] / 1000000);

            if (context.url) {
              // Somewhere a `<Redirect>` was rendered
              res.redirect(301, context.url);
            } else {
              // we're good, add in markup, send the response
              const direction    = getLanguageDirection(language);
              const cssDirection = direction === 'ltr' ? '' : '.rtl';

              store.dispatch(ssr.prepare());
              // console.log(require('util').inspect(store.getState(), { showHidden: true, depth: 2 }));
              const storeData = serialize(store.getState());

              const i18nData = serialize(
                {
                  initialLanguage: context.i18n.language,
                  initialI18nStore: pick(context.i18n.services.resourceStore.data, [
                    context.i18n.language,
                    context.i18n.options.fallbackLng,
                  ]),
                }
              );

              const rootDiv = `<div id="root" class="${direction}" style="direction: ${direction}">
  ${markup}
</div>
<script>
  window.__data = ${storeData};
  window.__i18n = ${i18nData};
</script>`;

              const html = htmlData
                .replace(/<html lang="en">/, `<html ${helmet.htmlAttributes.toString()} >`)
                .replace(/<title>.*<\/title>/, helmet.title.toString())
                .replace(/<\/head>/, `${helmet.meta.toString()}${helmet.link.toString()}${canonicalLink(req, language)}${alternateLinks(req, language)}${ogUrl(req, language)}</head>`)
                .replace(/<body>/, `<body ${helmet.bodyAttributes.toString()} >`)
                .replace(/semantic_v4.min.css/g, `semantic_v4${cssDirection}.min.css`)
                .replace(/<div id="root"><\/div>/, rootDiv);

              if (context.code) {
                res.status(context.code);
              }

              res.send(html);
            }
          })
          .catch(next);
      })
      .catch(next);
  });
}
