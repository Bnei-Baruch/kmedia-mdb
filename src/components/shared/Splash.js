import React from 'react';
import PropTypes from 'prop-types';
import { Header, Icon } from 'semantic-ui-react';

export const Splash = (props) => {
  const { text, subtext, icon, color, isLoading } = props;

  return (
    <Header as="h2" icon textAlign="center">
      <Icon name={icon} loading={isLoading} color={color} />
      <Header.Content>
        {text}
        {subtext ? <Header.Subheader>{subtext}</Header.Subheader> : null}
      </Header.Content>
    </Header>
  );
};

Splash.propTypes = {
  icon: PropTypes.string.isRequired,
  text: PropTypes.node.isRequired,
  subtext: PropTypes.node,
  color: PropTypes.string,
  isLoading: PropTypes.bool,
};

Splash.defaultProps = {
  text: '',
  subtext: null,
  color: 'black',
  isLoading: false,
};

export const LoadingSplash = props => (<Splash isLoading icon="spinner" {...props} />);
export const FrownSplash   = props => (<Splash icon="frown" color="orange" {...props} />);
export const ErrorSplash   = props => (<Splash icon="warning sign" color="red" {...props} />);
