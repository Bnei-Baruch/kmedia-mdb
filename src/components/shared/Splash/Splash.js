import React from 'react';
import PropTypes from 'prop-types';
import { Header, Icon } from 'semantic-ui-react';

export const Splash = props => {
  const { text, subtext = null, icon, color = 'black', isLoading = false, item = null } = props;

  return (
    <div className="splash">
      <Header icon as="h2" textAlign="center">
        <Icon name={icon} loading={isLoading} color={color} />
        <Header.Content>
          {text}
          {subtext ? <Header.Subheader data-testid='header.subHeader'>{subtext}</Header.Subheader> : null}
          {item ? <Header.Subheader>{item}</Header.Subheader> : null}
        </Header.Content>
      </Header>
    </div>
  );
};

Splash.propTypes = {
  icon: PropTypes.string.isRequired,
  text: PropTypes.node.isRequired,
  subtext: PropTypes.node,
  color: PropTypes.string,
  isLoading: PropTypes.bool,
  item: PropTypes.func,
};

export const LoadingSplash = props => (<Splash isLoading icon="circle notch" color="blue" {...props} />);
export const FrownSplash   = props => (<Splash icon="frown" color="orange" {...props} />);
export const ErrorSplash   = props => (<Splash icon="warning sign" color="red" {...props} />);
