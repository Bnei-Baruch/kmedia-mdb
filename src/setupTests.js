import React from 'react';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'react-testing-library/cleanup-after-each';
import 'jest-enzyme';
// import 'jest-axe/extend-expect'; // Testing the a11y
import 'jest-dom/extend-expect'; // adds custom jest matchers from jest-dom

import { BrowserRouter as Router } from 'react-router-dom';

configure({ adapter: new Adapter() });

global.requestAnimationFrame = cb => setTimeout(cb, 0);

export const shallowRender = (component, props) => shallow(React.createElement(component, props));

export const mountedRender = (component, props) => mount(React.createElement(component, props));

jest.mock('react-i18next', () => ({
  // this mock makes sure any components using the withNamespaces HoC receive the t function as a prop
  withNamespaces: () => (c) => {
    c.defaultProps = { ...c.defaultProps, t: k => k };
    return c;
  },
}));

export function mountWrapRouter(node) {
  return mount(<Router>{node}</Router>);
}

export function shallowWrapRouter(node) {
  return shallow(<Router>{node}</Router>);
}

