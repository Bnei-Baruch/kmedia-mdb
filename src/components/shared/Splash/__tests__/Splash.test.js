import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-enzyme';

import { Header, Icon } from 'semantic-ui-react';

import SplashFixture from '../__fixtures__/Splash.fixture';
import SplashWOSubtext from '../__fixtures__/SplashWOSubtext.fixture';
import { mountedRender, shallowRender } from '../../../../setupTests';

import { ErrorSplash, FrownSplash, LoadingSplash, Splash } from '../Splash';

[SplashFixture, SplashWOSubtext].forEach((fixture) => {
  describe('Splash', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mountedRender(Splash, { ...fixture.props });
    });

    it('renders error', () => {
      const { text, subtext } = fixture.props;
      const message           = `${text}${subtext || ''}`;
      expect(wrapper).toHaveText(message);
    });

    it('always render exactly one Header', () => {
      const headers = wrapper.find(Header);
      expect(headers).toHaveLength(1);
    });

    describe('the rendered Header', () => {
      it('always renders an Icon', () => {
        const icons = wrapper.find(Icon);
        expect(icons).toExist();
      });

      it('Icon has proper params', () => {
        const icon                             = wrapper.find(Icon);
        const { color, icon: name, isLoading: loading } = fixture.props;
        expect(icon).toHaveProp({ color, name, loading });
      });

      it('always renders an Header.Content', () => {
        const header = wrapper.find(Header.Content);
        expect(header).toExist();
      });
    });

    describe('subtext', () => {

      if (typeof fixture.props.subtext === 'undefined') {
        describe('when subtext is not supplied', () => {
          it('does not render Header.Subheader', () => {
            const subheader = wrapper.find(Header.Subheader);
            expect(subheader).not.toExist();
          });
        });
      } else {
        describe('when subtext is supplied', () => {
          it('renders Header.Subheader', () => {
            const subheader = wrapper.find(Header.Subheader);
            expect(subheader).toExist();
          });
          it('renders Header.Subheader with this subtext', () => {
            const subheader = wrapper.find(Header.Subheader);
            expect(subheader).toHaveText(fixture.props.subtext);
          });
        });
      }
    });

    describe('Renders the same as the last time', () => {
      it('Matches snapshot', () => {
        const tree = renderer.create(wrapper).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
});

describe('LoadingSplash', () => {
  const text          = 'Still loading...';
  const loadingSplash = shallowRender(LoadingSplash, { text });
  const tree1         = renderer.create(loadingSplash).toJSON();

  it('renders correctly', () => {
    const splash = shallowRender(Splash, { isLoading: true, icon: 'circle notch', color: 'blue', text });
    const tree2  = renderer.create(splash).toJSON();
    expect(tree1).toEqual(tree2);
  });

  it('matches snapshot', () => {
    expect(tree1).toMatchSnapshot();
  });
});

describe('FrownSplash', () => {
  const text        = 'Still loading...';
  const frownSplash = shallowRender(FrownSplash, { text });
  const tree1       = renderer.create(frownSplash).toJSON();

  it('renders correctly', () => {
    const splash = shallowRender(Splash, { icon: 'frown', color: 'orange', text });
    const tree2  = renderer.create(splash).toJSON();
    expect(tree1).toEqual(tree2);
  });

  it('matches snapshot', () => {
    expect(tree1).toMatchSnapshot();
  });
});

describe('ErrorSplash', () => {
  const text        = 'Nasty error...';
  const errorSplash = shallowRender(ErrorSplash, { text });
  const tree1       = renderer.create(errorSplash).toJSON();

  it('renders correctly', () => {
    const splash = shallowRender(Splash, { icon: 'warning sign', color: 'red', text });
    const tree2  = renderer.create(splash).toJSON();
    expect(tree1).toEqual(tree2);
  });

  it('matches snapshot', () => {
    expect(tree1).toMatchSnapshot();
  });
});
