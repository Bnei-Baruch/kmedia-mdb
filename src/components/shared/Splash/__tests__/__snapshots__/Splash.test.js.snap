// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorSplash matches snapshot 1`] = `
<div
  className="splash"
>
  <h2
    className="ui icon center aligned header"
  >
    <i
      aria-hidden="true"
      className="red warning sign icon"
    />
    <div
      className="content"
    >
      Nasty error...
    </div>
  </h2>
</div>
`;

exports[`FrownSplash matches snapshot 1`] = `
<div
  className="splash"
>
  <h2
    className="ui icon center aligned header"
  >
    <i
      aria-hidden="true"
      className="orange frown icon"
    />
    <div
      className="content"
    >
      Still loading...
    </div>
  </h2>
</div>
`;

exports[`LoadingSplash matches snapshot 1`] = `
<div
  className="splash"
>
  <h2
    className="ui icon center aligned header"
  >
    <i
      aria-hidden="true"
      className="blue circle notch loading icon"
    />
    <div
      className="content"
    >
      Still loading...
    </div>
  </h2>
</div>
`;

exports[`Splash Renders the same as the last time Matches snapshot 1`] = `
<div
  className="splash"
>
  <h2
    className="ui icon center aligned header"
  >
    <i
      aria-hidden="true"
      className="red warning sign icon"
    />
    <div
      className="content"
    >
      Server Error
      <div
        className="sub header"
      >
        No response from server
      </div>
    </div>
  </h2>
</div>
`;

exports[`Splash Renders the same as the last time Matches snapshot 2`] = `
<div
  className="splash"
>
  <h2
    className="ui icon center aligned header"
  >
    <i
      aria-hidden="true"
      className="red spinner loading icon"
    />
    <div
      className="content"
    >
      Server Error
    </div>
  </h2>
</div>
`;
